server/src/main/java/org/elasticsearch/action/admin/cluster/configuration/AddVotingConfigExclusionsRequest.java

```diff
@@ -56,32 +62,72 @@ public AddVotingConfigExclusionsRequest(String[] nodeDescriptions) {
      * @param nodeDescriptions Descriptions of the nodes whose exclusions to add - see {@link DiscoveryNodes#resolveNodes(String...)}.
      * @param timeout How long to wait for the added exclusions to take effect and be removed from the voting configuration.
      */
-    public AddVotingConfigExclusionsRequest(String[] nodeDescriptions, TimeValue timeout) {
+    public AddVotingConfigExclusionsRequest(String[] nodeDescriptions, String[] nodeIds, String[] nodeNames, TimeValue timeout) {
         if (timeout.compareTo(TimeValue.ZERO) < 0) {
             throw new IllegalArgumentException("timeout [" + timeout + "] must be non-negative");
         }
         this.nodeDescriptions = nodeDescriptions;
+        this.nodeIds = nodeIds;
+        this.nodeNames = nodeNames;
         this.timeout = timeout;
     }
 
     public AddVotingConfigExclusionsRequest(StreamInput in) throws IOException {
         super(in);
+        // TODO should this be removed in the latest version where nodeIds and nodeNames are used?
         nodeDescriptions = in.readStringArray();
+        // TODO which version to use here?
+        if (in.getVersion() == Version.V_EMPTY) {
+            nodeIds = in.readStringArray();
+            nodeNames = in.readStringArray();
+        }
         timeout = in.readTimeValue();
     }
 
     Set<VotingConfigExclusion> resolveVotingConfigExclusions(ClusterState currentState) {
         final DiscoveryNodes allNodes = currentState.nodes();
-        final Set<VotingConfigExclusion> resolvedNodes = Arrays.stream(allNodes.resolveNodes(nodeDescriptions))
+        Set<VotingConfigExclusion> allProcessedNodes = null;
+
+        if (nodeDescriptions.length >= 1) {
+            allProcessedNodes = Arrays.stream(allNodes.resolveNodes(nodeDescriptions)).map(allNodes::get)
+                                    .filter(DiscoveryNode::isMasterNode).map(VotingConfigExclusion::new).collect(Collectors.toSet());
+
+            if (allProcessedNodes.isEmpty()) {
+                throw new IllegalArgumentException("add voting config exclusions request for " + Arrays.asList(nodeDescriptions)
+                    + " matched no master-eligible nodes");
+            }
+        }
+        else {
+            DiscoveryNodes.NodeResolutionResults nodeResolutionResults;
+            Set<VotingConfigExclusion> resolvedNodes;
+            Set<VotingConfigExclusion> unresolvedNodes;
+
+            if (nodeIds.length >= 1) {
+                nodeResolutionResults = allNodes.resolveNodesExact(true, nodeIds);
+
+                unresolvedNodes = Arrays.stream(nodeResolutionResults.getUnresolvedNodes())
+                    .map(nodeId -> new VotingConfigExclusion(nodeId, "")).collect(Collectors.toSet());
```

Comment:
Suggest using a sentinel value other than `""` here to clarify why the node name is missing. I suspect this will save some confusion in a future debugging session.

```suggestion
                    .map(nodeId -> new VotingConfigExclusion(nodeId, "_absent_")).collect(Collectors.toSet());
```

Reply:
Done.